File: pdq/python/pdqhashing/types/hash256.py
#!/usr/bin/env python
# Copyright (c) Meta Platforms, Inc. and affiliates.

from random import randint

from pdqhashing.types.exceptions import PDQHashFormatException


class Hash256:
    """256-bit hashes with Hamming distance"""

    # 16 slots of 16 bits each.
    # See hashing/pdq/README-MIH.md in this repo for why not 8x32 or 32x8, etc.
    HASH256_NUM_SLOTS = 16

    HASH256_HEX_NUM_NYBBLES = 4 * HASH256_NUM_SLOTS

    def __init__(self) -> None:
        self.w = [0] * self.HASH256_NUM_SLOTS

    def getNumWords(self):
        return self.HASH256_NUM_SLOTS

    def clone(self):
        rv = Hash256()
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            rv.w[i] = self.w[i]
            i += 1
        return rv

    def __str__(self):
        i = self.HASH256_NUM_SLOTS - 1
        result = []
        while i >= 0:
            result.append("{:04x}".format(self.w[i] & 0xFFFF))
            i = i - 1
        return "".join(result)

    def __repr__(self):
        i = self.HASH256_NUM_SLOTS - 1
        result = []
        while i >= 0:
            result.append("{:04x}".format(self.w[i] & 0xFFFF))
            i = i - 1
        return "".join(result)

    def toHexString(self):
        return self.__str__()

    @classmethod
    def fromHexString(cls, s):
        if len(s) != cls.HASH256_HEX_NUM_NYBBLES:
            raise PDQHashFormatException("Incorrect length", s)

        rv = Hash256()
        i = cls.HASH256_NUM_SLOTS
        for x in range(0, len(s), 4):
            try:
                i -= 1
                rv.w[i] = int(s[x : x + 4], 16)
            except ValueError:
                raise PDQHashFormatException("Incorrect format", s)
        return rv

    @classmethod
    def hammingNorm16(cls, h):
        return cls.bitCount((int(h)) & 0xFFFF)

    @classmethod
    def bitCount(cls, x):
        x -= (x >> 1) & 0x55555555
        x = ((x >> 2) & 0x33333333) + (x & 0x33333333)
        x = ((x >> 4) + x) & 0x0F0F0F0F
        x += x >> 8
        x += x >> 16
        return x & 0x0000003F

    def clearAll(self):
        for i in range(self.HASH256_NUM_SLOTS):
            self.w[i] = 0

    def setAll(self):
        for i in range(self.HASH256_NUM_SLOTS):
            self.w[i] = 0xFFFF

    def hammingNorm(self):
        n = 0
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            n += self.hammingNorm16(self.w[i])
            i += 1
        return n

    def hammingDistance(self, that):
        n = 0
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            n += self.hammingNorm16(int((self.w[i] ^ that.w[i])))
            i += 1
        return n

    def hammingDistanceLE(self, that, d) -> bool:
        e = 0
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            e += self.hammingNorm16(int((self.w[i] ^ that.w[i])))
            if e > d:
                return False
            i += 1
        return True

    def setBit(self, k):
        self.w[(k & 255) >> 4] |= 1 << (k & 15)

    def flipBit(self, k):
        self.w[(k & 255) >> 4] ^= 1 << (k & 15)

    def bitwiseXOR(self, that):
        rv = Hash256()
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            rv.w[i] = int((self.w[i] ^ that.w[i]))
            i += 1
        return rv

    def bitwiseAND(self, that):
        rv = Hash256()
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            rv.w[i] = int((self.w[i] & that.w[i]))
            i += 1
        return rv

    def bitwiseOR(self, that):
        rv = Hash256()
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            rv.w[i] = int((self.w[i] | that.w[i]))
            i += 1
        return rv

    def bitwiseNOT(self):
        rv = Hash256()
        i = 0
        while i < self.HASH256_NUM_SLOTS:
            rv.w[i] = int((~self.w[i])) & 0xFFFF
            i += 1
        return rv

    def dumpBits(self):
        i = self.HASH256_NUM_SLOTS - 1
        str = []
        while i >= 0:
            word = self.w[i] & 0xFFFF
            j = 15
            bits = []
            while j >= 0:
                if (word & (1 << j)) != 0:
                    bits.append("1")
                else:
                    bits.append("0")
                j -= 1
            str.append(" ".join(bits))
            i -= 1
        return "\n".join(str)

    def dumpBitsAcross(self):
        i = self.HASH256_NUM_SLOTS - 1
        str = []
        while i >= 0:
            word = self.w[i] & 0xFFFF
            j = 15
            while j >= 0:
                if (word & (1 << j)) != 0:
                    str.append("1")
                else:
                    str.append("0")
                j -= 1
            i -= 1
        return " ".join(str)

    def dumpWords(self):
        return ",".join(str(v) for v in list(reversed(self.w)))

    def fuzz(self, numErrorBits):
        """Flips some number of bits randomly, with replacement.  (I.e. not all
        flipped bits are guaranteed to be in different positions; if you pass
        argument of 10 then maybe 2 bits will be flipped and flipped back, and
        only 6 flipped once.)"""
        rv = self.clone()
        i = 0
        while i < numErrorBits:
            rv.flipBit(randint(0, 255))
            i += 1
        return rv

    def __eq__(self, other) -> bool:
        if isinstance(other, (Hash256,)):
            for i in range(self.HASH256_NUM_SLOTS):
                if self.w[i] != other.w[i]:
                    return False
            return True
        else:
            return False

    def __gt__(self, other) -> bool:
        # pyre-fixme[16]: `int` has no attribute `__iter__`.
        for i in self.HASH256_NUM_SLOTS:
            if self.w[i] > other.w[i]:
                return True
            elif self.w[i] < other.w[i]:
                return False
        return False

    def __lt__(self, other) -> bool:
        # pyre-fixme[16]: `int` has no attribute `__iter__`.
        for i in self.HASH256_NUM_SLOTS:
            if self.w[i] < other.w[i]:
                return True
            elif self.w[i] > other.w[i]:
                return False
        return False

File: pdq/java/src/main/java/pdqhashing/types/Hash256.java
// ================================================================
// Copyright (c) Meta Platforms, Inc. and affiliates.
// ================================================================

package pdqhashing.types;

import pdqhashing.types.PDQHashFormatException;
import java.io.PrintStream;
import java.util.Arrays;
import java.util.Random;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 256-bit hashes with Hamming distance
 */
public class Hash256 implements Comparable<Hash256> {
  public static final int HASH256_NUM_SLOTS = 16;

  private static final String IO_FORMAT =
    "%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x%04x";

  private static final Pattern HEX_PATTERN = Pattern.compile(
    "([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})" +
    "([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})" +
    "([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})" +
    "([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})([0-9a-fA-F]{4})"
  );

  private static final Random RNG = new Random(); // for the fuzz() method

  // 16 slots of 16 bits each. See ../README.md for why not 8x32 or 32x8, etc.
  public final short[] w;

  // ----------------------------------------------------------------
  public Hash256() {
    this.w = new short[HASH256_NUM_SLOTS];
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      this.w[i] = 0;
    }
  }

  // ----------------------------------------------------------------
  public Hash256 clone() {
    Hash256 rv = new Hash256();
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      rv.w[i] = this.w[i];
    }
    return rv;
  }

  // ----------------------------------------------------------------
  public int getNumWords() {
    return HASH256_NUM_SLOTS;
  }

  // ----------------------------------------------------------------
  @Override
  public String toString() {
    return String.format(IO_FORMAT,
      (int)(this.w[15] & 0xffff), (int)(this.w[14] & 0xffff),
      (int)(this.w[13] & 0xffff), (int)(this.w[12] & 0xffff),
      (int)(this.w[11] & 0xffff), (int)(this.w[10] & 0xffff),
      (int)(this.w[9]  & 0xffff), (int)(this.w[8]  & 0xffff),
      (int)(this.w[7]  & 0xffff), (int)(this.w[6]  & 0xffff),
      (int)(this.w[5]  & 0xffff), (int)(this.w[4]  & 0xffff),
      (int)(this.w[3]  & 0xffff), (int)(this.w[2]  & 0xffff),
      (int)(this.w[1]  & 0xffff), (int)(this.w[0]  & 0xffff));
  }

  // ----------------------------------------------------------------
  public static Hash256 fromHexString(String s)
    throws PDQHashFormatException
  {
    Hash256 rv = new Hash256();

    Matcher matcher = HEX_PATTERN.matcher(s);
    if (matcher.find()) {
      for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
        String group = matcher.group(16 - i); // 1-up and reversed
        rv.w[i] = (short)(int)Integer.valueOf(group, 16);
      }
    } else {
      throw new PDQHashFormatException(s);
    }

    return rv;
  }

  // ----------------------------------------------------------------
  public static int hammingNorm16(short h) {
    return Integer.bitCount(((int)h) & 0xffff);
  }

  public void clearAll() {
    for (int i = 0; i < HASH256_NUM_SLOTS; i++)
      this.w[i] = 0;
  }

  public void setAll() {
    for (int i = 0; i < HASH256_NUM_SLOTS; i++)
      this.w[i] = ~0;
  }

  public int hammingNorm() {
    int n = 0;
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      n += hammingNorm16(this.w[i]);
    }
    return n;
  }

  public int hammingDistance(Hash256 that) {
    int n = 0;
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      n += hammingNorm16((short)(this.w[i] ^ that.w[i]));
    }
    return n;
  }

  public boolean hammingDistanceLE(Hash256 that, int d) {
    int e = 0;
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      e += hammingNorm16((short)(this.w[i] ^ that.w[i]));
      if (e > d)
        return false;
    }
    return true;
  }

  public void setBit(int k) {
    this.w[(k & 255) >> 4] |= 1 << (k & 15);
  }

  public void flipBit(int k) {
    this.w[(k & 255) >> 4] ^= 1 << (k & 15);
  }

  public Hash256 bitwiseXOR(Hash256 that) {
    Hash256 rv = new Hash256();
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      rv.w[i] = (short)(this.w[i] ^ that.w[i]);
    }
    return rv;
  }

  public Hash256 bitwiseAND(Hash256 that) {
    Hash256 rv = new Hash256();
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      rv.w[i] = (short)(this.w[i] & that.w[i]);
    }
    return rv;
  }

  public Hash256 bitwiseOR(Hash256 that) {
    Hash256 rv = new Hash256();
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      rv.w[i] = (short)(this.w[i] | that.w[i]);
    }
    return rv;
  }

  public Hash256 bitwiseNOT() {
    Hash256 rv = new Hash256();
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      rv.w[i] = (short)(~ this.w[i]);
    }
    return rv;
  }

  public void dumpBits(PrintStream o) {
    for (int i = HASH256_NUM_SLOTS - 1; i >= 0; i--) {
      int word = ((int)this.w[i]) & 0xffff;
      for (int j = 15; j >= 0; j--) {
        if ((word & (1 << j)) != 0) {
          o.printf(" 1");
        } else {
          o.printf(" 0");
        }
      }
      o.printf("\n");
    }
    o.printf("\n");
  }

  public void dumpBitsAcross(PrintStream o) {
    for (int i = HASH256_NUM_SLOTS - 1; i >= 0; i--) {
      int word = ((int)this.w[i]) & 0xffff;
      for (int j = 15; j >= 0; j--) {
        if ((word & (1 << j)) != 0) {
          o.printf(" 1");
        } else {
          o.printf(" 0");
        }
      }
    }
    o.printf("\n");
  }

  public void dumpWords(PrintStream o) {
    for (int i = HASH256_NUM_SLOTS - 1; i >= 0; i--) {
      short word = this.w[i];
      if (i < HASH256_NUM_SLOTS - 1)
        o.printf(" ");
      o.printf("%04x", ((int)word) & 0xffff);
    }
    o.printf("\n");
  }

  // Helper method to do sign extension
  private int toWord(short w) {
    return (int)w & 0xffff;
  }

  public void dumpInt64Vals(PrintStream o) {
    for (int i = HASH256_NUM_SLOTS - 1; i >= 0;) {
      long val = toWord(this.w[i--]);
      val = (val << 16) ^ toWord(this.w[i--]);
      val = (val << 16) ^ toWord(this.w[i--]);
      val = (val << 16) ^ toWord(this.w[i--]);
      o.printf("%d", val);
      if (i >= 0)
        o.printf(" ");
    }
    o.printf("\n");
  }

  /**
   * Flips some number of bits randomly, with replacement.  (I.e. not all
   * flipped bits are guaranteed to be in different positions; if you pass
   * argument of 10 then maybe 2 bits will be flipped and flipped back, and
   * only 6 flipped once.)
   */
  public Hash256 fuzz(int numErrorBits) {
    Hash256 rv = this.clone();
    for (int i = 0; i < numErrorBits; i++) {
      rv.flipBit(RNG.nextInt(256));
    }
    return rv;
  }

  // ================================================================
  // For Java collections: so we can make hashmaps of PDQ hashes, and so on.

  @Override
  public int hashCode() {
    return Arrays.hashCode(this.w);
  }

  @Override
  public boolean equals(Object othat) {
    if (othat instanceof Hash256) {
      Hash256 that = (Hash256)othat;
      for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
        if (this.w[i] != that.w[i]) {
          return false;
        }
      }
      return true;
    } else {
      return false;
    }
  }

  @Override
  public int compareTo(Hash256 that) {
    for (int i = 0; i < HASH256_NUM_SLOTS; i++) {
      if (this.w[i] < that.w[i]) {
        return -1;
      } else if (this.w[i] > that.w[i]) {
        return 1;
      }
    }
    return 0;
  }

};
